File: experiments/engine-kt/core/src/main/kotlin/com/garive/eng/kt/core/GovernedExecution.kt
package com.garive.eng.kt.core

import com.garive.eng.kt.tools.GovernedToolResult
import com.garive.eng.kt.tools.PreparationError
import com.garive.eng.kt.tools.PreparedToolCall
import com.garive.eng.kt.tools.ToolDefinition
import com.garive.eng.kt.tools.ToolIntent

/** Immutable exact tool capabilities frozen for one Kernel Execution. */
public data class AgentToolCapabilities(public val definitions: List<ToolDefinition>)

/** Governed result returned only after its required facts are durable. */
public data class CommittedGovernedResult(
    public val result: GovernedToolResult,
    public val throughPosition: ULong,
    public val suspensionBinding: GovernedSuspensionBinding?,
)

/** Runtime-owned durable identities carried into the Execution terminal. */
public sealed interface GovernedSuspensionBinding {
    /** One committed approval or external-input interaction. */
    public data class Interaction(
        public val suspensionId: String,
        public val interactionId: String,
        public val invocationId: String,
        public val preparedDigest: String,
    ) : GovernedSuspensionBinding

    /** One uncertain invocation requiring conclusive operator evidence. */
    public data class OperatorReconciliation(
        public val suspensionId: String,
        public val invocationId: String,
        public val preparedDigest: String,
    ) : GovernedSuspensionBinding
}

/** Runtime-owned durable authority and execution boundary used by the Agent loop. */
public interface GovernedEffectPort {
    /** Commits a C4 preparation rejection before returning model feedback. */
    public suspend fun reject(
        sourceModelRequestId: String,
        intent: ToolIntent,
        error: PreparationError,
    ): Result<CommittedGovernedResult>

    /** Allocates, authorizes, and executes or suspends one prepared call. */
    public suspend fun invoke(
        sourceModelRequestId: String,
        prepared: PreparedToolCall,
    ): Result<CommittedGovernedResult>
}

/** Runs the C0-C5 tool-capable bounded Agent loop. */
public suspend fun executeAgent(
    request: AgentTurnRequest,
    capabilities: AgentToolCapabilities,
    ports: AgentExecutionPorts,
    effects: GovernedEffectPort,
): ExecutionReport = executeKernel(request, ports, capabilities.definitions, effects)

File: engine/core/src/governed_execution.rs
//! Tool-capable execution ports layered above the C3 model-only boundary.

use std::{future::Future, pin::Pin};

use garive_tools::{
    GovernedToolResult, PreparationError, PreparedToolCall, ToolDefinition, ToolIntent,
};

use crate::{AgentExecutionPorts, AgentTurnRequest, ExecutionReport, PortFailure};

/// Immutable exact tool capabilities frozen for one Kernel Execution.
#[derive(Clone, Debug, PartialEq)]
pub struct AgentToolCapabilities {
    /// Full C4 definitions resolved into the Effective Agent Snapshot.
    pub definitions: Vec<ToolDefinition>,
}

/// Governed result returned only after its required facts are durable.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CommittedGovernedResult {
    /// Portable C5 result visible to Core.
    pub result: GovernedToolResult,
    /// Latest committed Session position after this result.
    pub through_position: u64,
}

/// Asynchronous result for one governed Runtime operation.
pub type GovernedEffectFuture<'a> =
    Pin<Box<dyn Future<Output = Result<CommittedGovernedResult, PortFailure>> + Send + 'a>>;

/// Runtime-owned durable authority/execution boundary used by the Agent loop.
pub trait GovernedEffectPort: Send {
    /// Commits a C4 preparation rejection before returning model feedback.
    fn reject<'a>(
        &'a mut self,
        source_model_request_id: &'a str,
        intent: &'a ToolIntent,
        error: &'a PreparationError,
    ) -> GovernedEffectFuture<'a>;

    /// Allocates, authorizes and executes or suspends one Prepared Call.
    fn invoke<'a>(
        &'a mut self,
        source_model_request_id: &'a str,
        prepared: &'a PreparedToolCall,
    ) -> GovernedEffectFuture<'a>;
}

/// Runs the C0-C5 tool-capable bounded Agent loop.
pub async fn execute_agent(
    request: &AgentTurnRequest,
    capabilities: &AgentToolCapabilities,
    ports: &mut AgentExecutionPorts<'_>,
    effects: &mut dyn GovernedEffectPort,
) -> ExecutionReport {
    crate::model_only::execute_with_tools(request, ports, &capabilities.definitions, effects).await
}
